gpflow/posteriors.py

```diff
@@ -242,20 +242,52 @@ def _conditional_with_precompute(
         Computes predictive mean and (co)variance at Xnew, *excluding* mean_function.
         Relies on cached alpha and Qinv.
         """
+
         Kmn = self.kernel(self.X_data, Xnew)
-        Knn = self.kernel(Xnew, full_cov=full_cov)
+        # compute kernel stuff
+        num_func = tf.shape(self.Y_data)[-1]  # R
+        N = tf.shape(Kmn)[-1]
+
+        # get the leading dims in Kmn to the front of the tensor Kmn
+        K = tf.rank(Kmn)
+        perm = tf.concat(
+            [
+                tf.reshape(tf.range(1, K - 1), [K - 2]),  # leading dims (...)
+                tf.reshape(0, [1]),  # [M]
+                tf.reshape(K - 1, [1]),
+            ],
+            0,
+        )  # [N]
+        Kmn = tf.transpose(Kmn, perm)  # [..., M, N]
+        leading_dims = tf.shape(Kmn)[:-2]
+
+        # get the leading dims in Knm to the front of the tensor Knm
+        perm_T = tf.concat(
+            [
+                tf.reshape(0, [1]),  # [M]
+                tf.reshape(K - 1, [1]),
+                tf.reshape(tf.range(1, K - 1), [K - 2]),  # leading dims (...)
+            ],
+            0,
+        )  # [N]
+        Knm = tf.transpose(Kmn, perm=perm_T)
 
+        Knn = self.kernel(Xnew, full_cov=full_cov)
         err = self.Y_data - self.mean_function(self.X_data)  # type: ignore
```

Comment:
Thanks for doing this @Andrew878 . I think some of this functionality could be replaced with the ` leading_transpose` method in `gpflow.utilities.ops`

Reply:
I was able to replace the `perm_T` with `leading_transpose` but I could not for the `perm` term. Happy to be shown a nice solution here (if possible)